### PR TITLE
server: If init fails, return 1 from main_common instead of calling exit(1).

### PR DESCRIPTION
**Pull request by rlazarus**
*Description*:
This change takes the `exit(1)` out of the Server::InstanceImpl constructor, and replaces it with a `return 1` from main_common by way of an `EnvoyException` from the constructor. The reason is that at Google we have some cleanup work to do after main_common returns (whether success or failure). If it calls exit() instead of returning, we never get the chance.

*Risk Level*: Low

*Testing*:
Updated unit tests to verify the exception is thrown in case of bad config. As a nice side benefit, we don't need death tests for those cases anymore.

*Docs Changes*:
N/A

*Release Notes*:
N/A

Signed-off-by: Reuven Lazarus <rzl@google.com>

### CODE REVIEW

File: source/server/server.cc [hunk 1]
```diff
@@ -71,7 +71,7 @@ InstanceImpl::InstanceImpl(Options& options, Network::Address::InstanceConstShar
               e.what());
     thread_local_.shutdownGlobalThreading();
     thread_local_.shutdownThread();
-    exit(1);
+    throw;
```
Comment (dnoe) [inline source/server/server.cc]: Can you add a header comment to the ctor saying it may throw EnvoyException ?

**Comment by mattklein123** [comment]
We can take this in 1.5.0.

File: source/exe/main_common.cc [hunk 2]
```diff
@@ -83,9 +83,13 @@ int main_common(OptionsImpl& options) {
...
-                              access_log_lock, component_factory, tls);
-  server.run();
+  try {
+    Server::InstanceImpl server(options, local_address, default_test_hooks, *restarter, stats_store,
+                                access_log_lock, component_factory, tls);
+    server.run();
+  } catch (const EnvoyException& e) {
+    return 1;
+  }
   ares_library_cleanup();
```
Comment (mattklein123) [inline source/exe/main_common.cc]: it doesn't really matter that much, but we should probably do this in all cases. I don't think it's worth fixing this though unless you want to.

Comment (rlazarus) [reply source/exe/main_common.cc]: Oh, true enough. Might as well fix it, stand by.

**Comment by mattklein123** [comment]
@rlazarus if you happen to still be around this needs a master merge and we can get it in for 1.5.0. Otherwise we can merge after. I'm going to tag in a little bit.

**Pull request merged**